### PR TITLE
Enable FP_SHARING for asset tracker and LTE BLE gateway

### PR DESCRIPTION
**Pull request by jtguggedal**
Mulitple threads use floats, so CONFIG_FP_SHARING needs to be
enabled for the applications to run.

### CODE REVIEW

**Comment by ioannisg** [comment]
I am wondering; where is CONFIG_FLOAT selected?

**Comment by jtguggedal** [comment]
> I am wondering; where is CONFIG_FLOAT selected?
> (Just checking, because CONFIG_FP_SHARING does not make sense without CONFIG_FLOAT)

It's selected by BSD lib: https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/master/lib/bsdlib/Kconfig#L11

**Comment by ioannisg** [comment]
And, in any case, we must have an IRQ handler for the FPU.

**Comment by jtguggedal** [comment]
> So, FP_SHARING is required to be enabled, if there are multiple Zephyr threads that make use of the Floating Point Services.
> 
> Instead of selecting CONFIG_FP_SHARING in the samples' prj.conf files, could we do that in the Kconfig sources of the actual modules?
> 
> I.e. have something like:
> 
> ```
> config FP_SHARING
>         default y if CONFIG_FLOAT
>         help 
>           Select FP_SHARING as this module makes use of the FP register
> ```
> 
> This depends, of course, on _some_ core module, e.g. the BSD lib selecting CONFIG_FLOAT=y.
> 
> @ulfalizer could you help here?

That change looks good to me, would be nice to not have to do it in projectconfigs. Perhaps even `default y if FLOAT && MULTITHREADING`.
Anyway, that change has to be done upstream and the NCS samples won't work until it's downstreamed. An option is to add this configs now, and remove them when the upmerge happens.

**Comment by lemrey** [comment]
I agree that it would be desirable to have a Kconfig module turn this on, instead of having to do that in the samples. In the meantime, I suppose we could accept this patch. Do you agree @ioannisg ?
@jtguggedal could you please take a look at CI?

**Comment by jtguggedal** [comment]
@lemrey Looks llike it's the nRF52 Desktop CI that's unhappy, while NCS CI passes. Reran it, but still fails. @wiminordic is off work, perhaps you have a clue what happens, @pdunaj ?
Anyway, this patch doesn't touch desktop, so I think we're safe.

**Comment by ioannisg** [comment]
> 
> 
> I agree that it would be desirable to have a Kconfig module turn this on, instead of having to do that in the samples. In the meantime, I suppose we could accept this patch. Do you agree @ioannisg ?
> @jtguggedal could you please take a look at CI?

No, @lemrey I do not agree, but, perhaps I haven't explained myself clear enough :)

So, let's trust for now that CONFIG_FLOAT is always enabled in the BSD library. And BSD lib is always compiled with any application on nrf9160.

So, a next-best solution (compared to the ideal of auto-enabling these switches) is to select FP_SHARING (if FLOAT) in the libraries that implement threads that use floating point services. In our case, the sensor and gps libraries. Isn't this ok for you?

**Comment by jtguggedal** [comment]
@ioannisg @lemrey I've updated the PR accordingly.

**Comment by lemrey** [comment]
`applications/asset_tracker/src/orientation_detector/orientation_detector.c` needs it as well, but it's not a component. Perhaps we could keep the option on in the sample that uses it.

File: drivers/gps_sim/Kconfig [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 menuconfig GPS_SIM
 	bool "GPS simulator"
+	imply FP_SHARING
```
Comment (ioannisg) [inline drivers/gps_sim/Kconfig]: Looks good; should we also add if FLOAT here? Cause FP_SHARING has no meaning without it.
Same in the SENSOR_SIM below

Comment (jtguggedal) [reply drivers/gps_sim/Kconfig]: True, thanks!

**Comment by jtguggedal** [comment]
@lemrey @ioannisg You're right, the asset tracker and gateway would fail if the simulators were both enabled, so had to bring back the project configs selecting FLOAT and FP_SHARING.

**Comment by jtguggedal** [comment]
Thanks for the explanations and link, @ulfalizer ! I definitely need to read up on Kconfig best practices.
The GPS and sensor simulators use `double` type, but as far as I understand, the compiler would do some magic for float divisions even if `FLOAT` is not enabled. The drivers should work without FPU enabled. So perhaps the solution would to `select FP_SHARING if FLOAT` ?

@ioannisg @lemrey I'm starting to wonder if we should just `select FP_SHARING` in BSD lib's Kconfig as that's where `FLOAT` is selected in the first place?

**Comment by ioannisg** [comment]
> 
> 
> Does `GPS_SIM` require `FLOAT` and `FP_SHARING`? In that case, `imply` isn't the right solution, because with `imply`, it's still possible to disable `FLOAT` and `FP_SHARING`.
> 
> It's a good idea to always think of Kconfig changes in terms of going into the `menuconfig` and changing things, and making sure that no "bad" configurations can be created that way. Even if you never use the `menuconfig` (I know that many people never do :), it will automatically lead to a good dependency structure. Can think of defconfig files as "saved sessions" in `menuconfig`.
> 
> If `GPS_SIM` requires those symbols to be enabled, then just having `depends on` and enabling `FLOAT` and `FP_SHARING` in the defconfig files might not be so bad. Bit more spam, but simple and straightforward.
> 
> If you go with `select`, be careful with dependencies. See the [select section](https://docs.zephyrproject.org/latest/guides/kconfig/index.html#select-statements) on the Kconfig best practices page.

@ulfalizer thanks for helping out.
The situation is as follows:
- **each** module that expects to use the floating point services requires to  `select FLOAT`
- if **more than one** modules expect to use the floating point services, the project application requires `FP_SHARING`
How can we achieve this?

**Comment by ulfalizer** [comment]
@ioannisg 
No way to do arithmetic in Kconfig, so would have to do something like for three modules:

```
config FORCE_FP_SHARING
        def_bool (A && B) || (A && C) || (B && C)
        select FP_SHARING
```

Not really scalable. Maybe Kconfig isn't the best place to handle it.

Perhaps the logical meaning of ``FP_SHARING`` could be changed so that FP sharing is turned on if required, even if ``FP_SHARING`` is `n`. It'd just be a way to force it.

**Comment by jtguggedal** [comment]
@ioannisg How do you want me to proceed with this PR? The simplest solution that I see, is to let BSD lib's Kconfig `select FP_SHARING`, as that's the one selecting `FLOAT` in the first place.

**Comment by ioannisg** [comment]
There does not seem to be an ideal solution @jtguggedal here, as discussed with @ulfalizer.

I am leaning towards the following:
- Is there any application/sample/project for nRF9160 that does _not_ require the BSD lib? Assuming that there is _not_, and since BSD lib requires CONFIG_FLOAT, let it be this way! I.e. let BSD lib define CONFIG_FLOAT=y.
- On the other hand, (I can imagine that) it is possible that there are applications/samples/projects which do not link-in any modules that require floating-point. E.g. an application without GPS, without sensors, etc. So we should _not_ enable FP_SHARING in BSD lib. FP_SHARING is not a free Kconfig option; it comes with a price of flash/ram/run-time overhead, so it should be avoided if there is just a single thread using the FP context.
- And then, each module that implements threads which use the FP context, shall, then, select FP_SHARING. 

does this make sense, @jtguggedal @ulfalizer ?

**Comment by jtguggedal** [comment]
@ioannisg That's fine by me. Let's have the samples / apps enable it. I'll update the PR when @ulfalizer gives the green light.

**Comment by ulfalizer** [comment]
Sounds fine to me. Probably best to set `CONFIG_FLOAT=y` and `CONFIG_FP_SHARING=y` as needed in `prj.conf`, since these are configuration options for the applications proper, and not fixed hardware settings.

`prj.conf` modifies the configuration, while `Kconfig.defconfig` modifies the Kconfig files (which determine the valid configurations for the machine).

**Comment by ioannisg** [comment]
> 
> 
> Sounds fine to me. Probably best to set `CONFIG_FLOAT=y` and `CONFIG_FP_SHARING=y` as needed in `prj.conf`, since these are configuration options for the applications proper, and not fixed hardware settings.
> 
> `prj.conf` modifies the configuration, while `Kconfig.defconfig` modifies the Kconfig files (which determine the valid configurations for the machine).

If we add then in the prj.conf files, we will need to be adding text, explaining _why_ these are needed.
If we add them in the Kconfig of modules, this is somehow implicit.

AFAIK, we don't just add HW configuration in Kconfig files, we also add software configuration options. 

Anyway, its' not so important where we add them, as long they are added :)

**Comment by ulfalizer** [comment]
@ioannisg 
Adding some comments to `prj.conf` seems like a good solution to me.

Think we should strive to describe configuration options in Kconfig files, and describe the configuration itself in configuration files. The current mishmash is pretty confusing.

**Comment by lemrey** [comment]
There are pros and cons in both options, but I would go with letting the libraries set `FP_SHARING` themselves, instead of relying on the user to do so in the project files. The reason is that if the user forgets to do so, or if they are simply unaware of how a specific library is implemented, their program will hit a hard-fault; that's not nice.

I am *very* sure that this will be often the case.

Not to mention that at the same time we'd have to maintain that setting in several `prj.conf` files (where it will need to be maintained), and for sure it is going to be copy-pasted in projects where it is not necessary, eventually..

If pulling in two libraries results by default in a hard-fault, unless a specific option is turned on, just doesn't seem like the most user-friendly solution. Let this work out of the box for most people, and let the rest turn it off where appropriate.

**Comment by jtguggedal** [comment]
@lemrey @ioannisg How about setting it both in the drivers with `select FP_SHARING if FLOAT`, and also in the projects' `prj.conf` with comments explaining why `FP_SHARING` is also enabled. For example:
```
CONFIG_FLOAT=y
# Required because multiple threads use floating-point operations
CONFIG_FP_SHARING=y
```

I think it's useful to have it both places to avoid users experiencing strange hard-faults when using the simulators, and also serve as an example that it's necessary for all projects that use floats in multiple contexts, not only the simulators.

**Comment by ioannisg** [comment]
> 
> 
> @lemrey @ioannisg How about setting it both in the drivers with `select FP_SHARING if FLOAT`, and also in the projects' `prj.conf` with comments explaining why `FP_SHARING` is also enabled. For example:
> 
> ```
> CONFIG_FLOAT=y
> # Required because multiple threads use floating-point operations
> CONFIG_FP_SHARING=y
> ```
> 
> I think it's useful to have it both places to avoid users experiencing strange hard-faults when using the simulators, and also serve as an example that it's necessary for all projects that use floats in multiple contexts, not only the simulators.
> 
> Update: I fixed the commits per my suggestion above. Please let me know if (and how) I should change them further.

@jtguggedal duplicate definitions are bad, IMO because: if someone re-implements the library at some time, and does not need the FP_SHARING any more, they need to go into all the projects that use this library and remove the Kconfig symbol. Not good.

As I stressed above, we need to enable FP_SHARING only when we are obliged to, otherwise, we introduce a cost for no reason.

I think we discussed it thoroughly in this PR. I am also in favor of adding the configuration in the libraries that are known that they use the Floating-Poing context, and skip from the application configurations as @lemrey suggested. If there's an application that directly uses the fp (i.e. not via a library) of course we add the options in the respective prj.conf.

Also, you can definitely add the explanatory comments, as well, when you select FP_SHARING and select FLOAT

**Comment by jtguggedal** [comment]
@ioannisg  Ok, I'll remove it from the prj.conf if @lemrey also agrees.
We just need to be aware that if anyone disables the simulators (as will probably be the case when GPS is available), the app will fail again because the `orientation_detector`, called from system wokqueue, also uses FPU when it's enabled. I'm not sure how obvious it is to users that the missing `CONFIG_FP_SHARING=y` is the reason for the hardfault they get.

**Comment by lemrey** [comment]
@jtguggedal Me and @ioannisg agree that libraries shall turn on `FLOAT` and `FP_SHARING` as necessary, including `bsdlib` since socket calls can happen from any number of threads.
This way, the `asset_tracker` should work fine even with the simulators off, right?

**Comment by jtguggedal** [comment]
@lemrey @ioannisg Alright, then I think it should be okay. GPS sim, sensor sim and bsdlib selects FP_SHARING if FLOAT is enabled. Sorry for all the misunderstandings, hopefully it should be good now :)

**Comment by ioannisg** [comment]
A bit out of scope, but I am wondering now, @lemrey : is the BSD lib compiled with FP_SHARING? How does this work, now, if it is not?

**Comment by lemrey** [comment]
Thank you @jtguggedal for your patience and thanks @ioannisg for weighing in on this.

**Pull request merged**